Threads - Untrusted Teardown + NapDtor

rennergade opened the pull request:
This PR combines the teardown handler (which was already looked over mostly) and functionality to properly free the NaClApp structs.

I've removed the signal teardown for trusted code since we should really abort if anything happens there now that we are checking vmmap protections for wonky addresses.

The napdtor includes switching to shutting down desc's through NaCl on exit, and handling situations where a cage becomes an orphan, as well as creating the Dtor function which is triggerd in NaClAppThreadTeardown.

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
try3

rennergade committed:
PR changes

rennergade committed:
PR changes

rennergade committed:
PR changes

rennergade committed:
PR changes

rennergade committed:
PR changes

rennergade committed:
PR changes

rennergade committed:
PR changes

rennergade committed:
Merge branch 'develop' into percage-signals3

rennergade committed:
add nap dtor

rennergade committed:
add nap dtor

rennergade committed:
add nap dtor

rennergade committed:
add nap dtor

rennergade committed:
add nap dtor

rennergade committed:
add nap dtor

rennergade committed:
add nap dtor

rennergade committed:
add nap dtor

rennergade committed:
add nap dtor

rennergade committed:
add nap dtor

rennergade committed:
add nap dtor

rennergade committed:
add nap dtor

rennergade committed:
add nap dtor

rennergade committed:
add nap dtor

rennergade committed:
testing nap free functions

rennergade committed:
testing nap free functions

rennergade committed:
testing nap free functions

rennergade committed:
testing nap free functions

rennergade committed:
testing nap free functions

rennergade committed:
testing nap free functions

rennergade committed:
testing nap free functions

rennergade committed:
testing nap free functions

rennergade committed:
testing nap free functions

rennergade committed:
testing nap free functions

rennergade committed:
testing nap free functions

rennergade committed:
testing nap free functions

rennergade committed:
testing nap free functions

rennergade committed:
testing nap free functions

rennergade committed:
testing nap free functions

rennergade committed:
testing nap free functions

rennergade committed:
testing nap free functions

rennergade committed:
testing nap free functions

rennergade committed:
testing nap free functions

rennergade committed:
testing nap free functions

rennergade committed:
testing nap free functions

rennergade committed:
testing nap free functions

rennergade committed:
testing nap free functions

rennergade committed:
testing nap free functions

rennergade committed:
testing nap free functions

rennergade committed:
testing nap free functions

rennergade committed:
testing nap free functions

rennergade committed:
testing nap free functions

rennergade committed:
trusted fault test

rennergade committed:
trusted fault test

rennergade committed:
trusted fault test

rennergade committed:
Merge branch 'develop' into nap-free

rennergade committed:
revert to abort on trusted faults

rennergade committed:
fix tls

rennergade committed:
trusted fault test

rennergade committed:
trusted fault test

rennergade committed:
finish nap free

rennergade committed:
finish nap free

rennergade committed:
vmmap

rennergade committed:
vmmap

rennergade committed:
vmmap

rennergade committed:
vmmap

rennergade committed:
vmmap

rennergade committed:
vmmap

rennergade committed:
vmmap

rennergade committed:
vmmap tests

rennergade committed:
vmmap tests

rennergade committed:
vmmap tests

rennergade committed:
vmmap tests

rennergade committed:
vmmap tests

rennergade committed:
vmmap tests

rennergade committed:
vmmap tests

rennergade committed:
vmmap tests

rennergade committed:
vmmap tests

rennergade committed:
cleanup

jesings commented on src/trusted/service_runtime/nacl_app_thread.c:
```diff
@@ -208,6 +217,34 @@ struct NaClApp *NaClChildNapCtor(struct NaClApp *nap, int child_cage_id, enum Na
+void NaClAppCloseFDs(struct NaClApp *nap) {
+  struct NaClDesc *ndp = NULL;
+  int             fd = 0;
+  int             i;
```
There's no reason not to declare i within the while loop given that it's a loop variable and it's not used after the loop

jesings commented on src/trusted/service_runtime/nacl_app_thread.c:
```diff
@@ -208,6 +217,34 @@ struct NaClApp *NaClChildNapCtor(struct NaClApp *nap, int child_cage_id, enum Na
 
+void NaClAppCloseFDs(struct NaClApp *nap) {
+  struct NaClDesc *ndp = NULL;
+  int             fd = 0;
```
This assignment to fd is never used, as it's reassigned at the start of the loop

jesings commented on src/trusted/service_runtime/nacl_app_thread.c:
```diff
@@ -208,6 +217,34 @@ struct NaClApp *NaClChildNapCtor(struct NaClApp *nap, int child_cage_id, enum Na
 }
 
+void NaClAppCloseFDs(struct NaClApp *nap) {
+  struct NaClDesc *ndp = NULL;
```
Ndp can be declared at the top of the innermost scope it will be used in, but that's just preference.

jesings commented on src/trusted/service_runtime/nacl_app_thread.c:
```diff
@@ -364,14 +391,45 @@ void NaClAppThreadTeardown(struct NaClAppThread *natp) {
+  }
 
+  /* Sadly we've created orphans :'(
+   * Remove parent from any children, and hope they don't become the Batman
```
Haha

jesings commented on src/trusted/service_runtime/nacl_app_thread.c:
```diff
@@ -364,14 +391,45 @@ void NaClAppThreadTeardown(struct NaClAppThread *natp) {
+   * Remove parent from any children, and hope they don't become the Batman
+   */
+  NaClXMutexLock(&nap->children_mu);
+  for (int i = 0; i < (&nap->children)->ptr_array_space; i++) {
```
Isn't ptr_array_space not the right field (i.e. the number of slots allocated rather than the number of slots used?

rennergade replied:
There isn't really functionality to just go over the slots used so I have to run over all that are allocated. This is still a small number.

jesings replied:
The num_entries field should do just that though?

jesings commented on src/trusted/service_runtime/nacl_app_thread.c:
```diff
@@ -653,6 +744,17 @@ int NaClAppThreadSpawn(struct NaClAppThread     *natp_parent,
+  } 
+  else {
+    natp_child->is_cage_mainthread = false;
+    natp_child->cage_mainthread = natp_parent;
```
Wait is this always true?
I'd think you should probably just have the natp for the cage mainthread refer to itself with the cage_mainthread field and then you can set `natp_child->cage_mainthread = natp_parent->cage_mainthread` here.

jesings commented on src/trusted/service_runtime/nacl_app_thread.c:
```diff
@@ -710,6 +812,113 @@ void NaClAppThreadDelete(struct NaClAppThread *natp) {
+ * a faulted thread is caught in the signal handler (nacl_signal.c)
+ * 
+ * The fault teardown function will cycle through all child threads in cage,
+ * tear them down, and call pthread_cancle upon them
```
Typo

jesings commented on src/trusted/service_runtime/nacl_app_thread.c:
```diff
@@ -710,6 +812,113 @@ void NaClAppThreadDelete(struct NaClAppThread *natp) {
+}
+
+
+void AddToFatalThreadTeardown(struct NaClAppThread *natp) {
+    if (natp_to_teardown == natp) return;
+    NaClXMutexLock(&teardown_mutex);
+    natp_to_teardown = natp;
+    natp->tearing_down = true;
+    NaClXCondVarBroadcast(&reapercv);
+    NaClXMutexUnlock(&teardown_mutex);
+}
+
```
Hmm how come this doesn't just let anyone clobber over the values in this field whenever they want after they're both assigned?

jesings commented on src/trusted/service_runtime/nacl_signal_common.c:
```diff
@@ -134,7 +138,23 @@ void NaClSignalHandleUntrusted(int signal,
+      if (!natp->cage_mainthread->tearing_down) AddToFatalThreadTeardown(natp->cage_mainthread);
+    }
+    // wait here while we get cleaned up
+    while (1);
```
At this point you should probably do a `pause()` instead of busy loop?

jesings replied:
That is the purpose of pause after all

rennergade replied:
I'm almost certain I can't use pause() in a signal handler. My capabilities here are pretty minimal.

jesings commented on src/trusted/service_runtime/sel_ldr.c:
```diff
@@ -310,44 +318,83 @@ int NaClAppWithSyscallTableCtor(struct NaClApp               *nap,
   return NaClAppWithSyscallTableCtor(nap, nacl_syscall);
 }
 
+void NaClAppDtor(struct NaClApp *nap) {
```
This is much better than whatever garbage we were doing before.

jesings reviewed (commented):
I'm not sure what happens if the main thread calls pthread_exit but there are still other threads in the process.
Overall this is awesome stuff.

rennergade committed:
cleanup

rennergade committed:
cleanup

jesings commented on src/trusted/service_runtime/nacl_signal_common.c:
```diff
@@ -134,7 +138,23 @@ void NaClSignalHandleUntrusted(int signal,
+      if (!natp->cage_mainthread->tearing_down) AddToFatalThreadTeardown(natp->cage_mainthread);
+    }
+    // wait here while we get cleaned up
+    while (1);
```
Oh right makes sense

rennergade committed:
cleanup

rennergade commented:
Going to close this for a bit while I fix it up.

rennergade committed:
cleanup

rennergade committed:
cleanup

rennergade committed:
cleanup

rennergade committed:
cleanup

rennergade committed:
cleanup

rennergade committed:
cleanup

rennergade committed:
cleanup

rennergade committed:
cleanup

rennergade commented:
@jesings I think I addressed your concerns here.

rennergade commented:
fork_simple is being wonky in the test suite, but @kapkic is fixing that.

rennergade committed:
cleanup

rennergade committed:
cleanup

rennergade committed:
cleanup

rennergade committed:
Fork locking

rennergade commented:
Dealing with that pesky fork_simple test actually exposed a bug in how we're duplicating nap information on fork/exec. Closing this until I fix that bug.

rennergade committed:
fork locking

rennergade committed:
fork locking

rennergade committed:
fork locking

rennergade committed:
nap ctor copying

rennergade committed:
nap ctor copying

rennergade committed:
nap ctor copying

rennergade committed:
nap ctor copying

rennergade committed:
nap ctor copying

rennergade committed:
nap ctor copying

rennergade committed:
nap ctor copying

rennergade committed:
nap ctor copying

rennergade committed:
nap ctor copying

rennergade committed:
nap free

rennergade committed:
nap free

rennergade committed:
nap free

rennergade committed:
nap free

rennergade committed:
final comment fixes

rennergade commented:
Finally got that green check mark.

I think this is good to go now. We changed the failing test (fork_simple) to a test that isn't insane. But regardless I fixed what was breaking the original one, which was a problem with the service threads once I started freeing the Nap structs properly.

I removed the service threads entirely, we don't use them at all. So this now makes things a bit cleaner to look at in gdb as a positive side effect (as well as being faster? probably negligibly but still...)

jesings commented on src/trusted/service_runtime/nacl_app_thread.c:
```diff
@@ -65,9 +73,14 @@ struct NaClApp *NaClChildNapCtor(struct NaClApp *nap, int child_cage_id, enum Na
+    nap_child->argv = calloc((nap_child->argc + 1), sizeof(char*));
+    for (int i = 0; i < nap_child->argc; i++) nap_child->argv[i] = strdup(nap_parent->argv[i]);
+    nap_child->binary = strdup(nap_parent->binary);
+  }
```
I'd rather this part be after all of the other nap_child assigns just to make the function better organized.

jesings commented on src/trusted/service_runtime/nacl_app_thread.c:
```diff
@@ -147,14 +159,8 @@ struct NaClApp *NaClChildNapCtor(struct NaClApp *nap, int child_cage_id, enum Na
-
   
   /* duplicate file descriptor table starting at child_fd = 3 (0-2 setup previously)*/
-  NaClXMutexLock(&nap_parent->mu);
```
Do we not want to lock this to make sure there are no race conditions with the parent's fd_cage_table?

rennergade replied:
I moved this so the mutex is actually held outside of the entire ChildCtor function.

jesings commented on src/trusted/service_runtime/nacl_app_thread.c:
```diff
@@ -398,42 +454,71 @@ void NaClAppThreadTeardown(struct NaClAppThread *natp) {
-  NaClSignalStackUnregister();
+
+  // again we trigger the lock outside for non-active threads, we also cant unregister the sig stack
+  // here so lets destroy everything
```
This comment doesn't make sense right before you unregister the signal stack...

jesings commented on src/trusted/service_runtime/nacl_app_thread.c:
```diff
@@ -398,42 +454,71 @@ void NaClAppThreadTeardown(struct NaClAppThread *natp) {
+
+
+  if (natp->is_cage_mainthread) {
+    // we have to wait for NaClWaitForMainThreadToExit on Main/Exec
```
Why for exec? (not sure I understand 100% what's being waited on).

jesings commented on src/trusted/service_runtime/nacl_app_thread.c:
```diff
@@ -398,42 +454,71 @@ void NaClAppThreadTeardown(struct NaClAppThread *natp) {
+    NaClLog(3, " NaClThreadExit\n");
+    NaClThreadExit();
+    NaClLog(LOG_FATAL, "NaClAppThreadTeardown: NaClThreadExit() should not return\n");
+    /* NOTREACHED */
```
I don't think this comment is necessary

jesings replied:
Even if it was there previously.

jesings commented on src/trusted/service_runtime/nacl_app_thread.c:
```diff
@@ -710,6 +806,114 @@ void NaClAppThreadDelete(struct NaClAppThread *natp) {
+}
+
+
+void AddToFatalThreadTeardown(struct NaClAppThread *natp) {
```
This mutex locking architecture still seems highly problematic to me given that it's undefined behavior for a thread to unlock a mutex that was locked by another thread.

jesings replied:
From what I understand, this function is just called by any old thread, and the reaper thread is responsible for unlocking in FatalThreadTeardown.

rennergade replied:
I forgot that was the case. Makes sense. Will replace with a CV.

jesings commented on src/trusted/service_runtime/nacl_app_thread.c:
```diff
@@ -710,6 +806,114 @@ void NaClAppThreadDelete(struct NaClAppThread *natp) {
+    NaClAppThreadTeardownInner(natp_to_teardown, false);
+    NaClThreadCancel(thread);
+    natp_to_teardown = NULL;
+    NaClXMutexUnlock(&teardown_mutex);
```
weird indentation

jesings commented on src/trusted/service_runtime/nacl_app_thread.c:
```diff
@@ -710,6 +806,114 @@ void NaClAppThreadDelete(struct NaClAppThread *natp) {
+}
+
+void ThreadReaper(void* arg) {
+  while (reap) {
+    NaClXMutexLock(&reapermut);
+
+    NaClXCondVarWait(&reapercv, &reapermut);
+    if (reap) FatalThreadTeardown();
+
+    NaClXMutexUnlock(&reapermut);
+  }
```
Is there a reason to have the locking/unlocking in the loop, rather than outside? You get a signal when reap is set to false, so this wouldn't change anything really.

jesings commented on src/trusted/service_runtime/nacl_app_thread.c:
```diff
@@ -710,6 +806,114 @@ void NaClAppThreadDelete(struct NaClAppThread *natp) {
+
+void DestroyReaper(void) {
+  reap = false;
+  NaClXCondVarBroadcast(&reapercv);
```
For a single thread, that can recieve a signal on a cv, you should use signal instead of broadcast.

jesings commented on src/trusted/service_runtime/nacl_syscall_common.c:
```diff
@@ -381,13 +381,20 @@ int32_t NaClSysGetppid(struct NaClAppThread *natp) {
-
   struct NaClApp *nap = natp->nap;
+
+  /* to close a cage we need to unref the vmmap before fofficially
```
Typo

rennergade committed:
final comment fixes

jesings reviewed (changes requested):
Sorry, I still think the mutex thing is not quite right. Overall this is looking much better though.

rennergade committed:
mutex fixes

rennergade committed:
mutex fixes

rennergade committed:
exec binary

rennergade committed:
exec binary

rennergade committed:
exec binary

rennergade committed:
exec binary

rennergade committed:
exec binary

rennergade committed:
exec binary

rennergade committed:
exec binary

rennergade committed:
exec binary

rennergade commented:
Alright, I think this is good now. @jesings I made the changes you requested and changed the arg translation a bit for fork/exec so they could be freed appropriately.

rennergade committed:
final comments, part 2?

rennergade committed:
final comments, part 2?

jesings reviewed (approved):
Looks good!
